torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -131,13 +139,54 @@ void DistEngine::computeDependencies(
             //  its ancestors need to be executed as well.
             if (dynamic_cast<RecvRpcBackward*>(nextFn)) {
               recvBackwardEdges.emplace_back(edge);
+            } else if (
+                auto accumulateGradFn = dynamic_cast<AccumulateGrad*>(nextFn)) {
+              TORCH_INTERNAL_ASSERT(
+                  !accumulateGradFn->variable.defined() ||
+                      nextFn ==
+                          torch::autograd::impl::try_get_grad_accumulator(
+                              accumulateGradFn->variable)
+                              .get(),
+                  "The AccumulateGrad point in the variable should be the "
+                  "same as the one in autograph");
+              // Replace an AccumulateGrad node by DistAccumulateGrad,
+              // because the former accumulates grads to the variable's '.grad'
+              // without considering the context id. That may cause data race
+              // on the '.grad', since multiple context ids may have grads for
+              // the same '.grad'.
+              auto distAccumulateGradFn = std::make_shared<DistAccumulateGrad>(
+                  // The pointer to 'AccumulateGrad' in
+                  // 'accumulateGradFn->variable' will be replaced by a pointer
+                  // to 'distAccumulateGradFn'.
+                  std::move(*accumulateGradFn),
```

Comment:
Why do we need to move the `AccumulateGrad` node here? Can we just keep a copy the `accumulateGradFn ` `shared_ptr` (assuming we can get shared_ptr from `edge.function`)? This will keep the existing `AccumulateGrad` node intact?